### PR TITLE
CLUe performance

### PR DESCRIPTION
**Pull request by Dantemss**
- Added StudentBookPart and PeriodBookPart models
- Cached student and period CLUes in StudentBookPart and PeriodBookPart when an immediate feedback question is answered or feedback becomes available
- Read CLUes from StudentBookPart and PeriodBookPart in the FakeClient
- Recorded whether the book_part_uuid belongs to a page or chapter and use only page CLUes in practice worst areas to avoid duplicate exercises.
- Specs

### CODE REVIEW

**Comment by Dantemss** [comment]
These new caches should have all the fields we need in the student/teacher guides (when combined with the book structure from the ecosystem itself). So once we get rid of the trouble flags we can remove the old TaskCache and PeriodCache tables.

File: db/background_migrate/20200501125646_initialize_clue_caches.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            period: role.student.period, book_part_uuid: book_part_uuid, is_page: false
+          ) if role.student?
+        end
+      end
+    end
+  end
+
+  def down
+  end
+end
```
Comment (nathanstitt) [inline db/background_migrate/20200501125646_initialize_clue_caches.rb:49]: this seems like it'll take a long time to compute?

File: db/migrate/20200428154625_add_parent_book_part_uuid_to_content_pages.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  def change
+    add_column :content_pages, :parent_book_part_uuid, :uuid
+
+    reversible do |dir|
+      dir.up do
+        Content::Models::Book.find_each do |book|
+          update_tree_and_pages! book.tree, book.uuid
+
+          book.save!
+        end
```
Comment (nathanstitt) [inline db/migrate/20200428154625_add_parent_book_part_uuid_to_content_pages.rb:11]: what changes on the book that causes it to need saved?

File: lib/tutor/subsystems/association_extensions.rb [hunk 3]
```diff
@@ -31,20 +31,24 @@ def set_subsystem_options(association_name, scope, options, is_belongs_to=false)
...
+        return if ['none', 'ignore'].include?(subsystem_name)
 
         # Find the string before the first ::, thats the subsystem module
-        my_subsystem_name = self.name.to_s[/(\w+?)::/,1].underscore
+        my_subsystem_name = self.name.to_s[/(\w+?)::/, 1]&.underscore
 
-        # Only extend subsystem's that are listed as valid.
-        return unless Tutor::SubSystems.valid_name?(my_subsystem_name)
+        # Only extend subsystems that are listed as valid.
+        return unless my_subsystem_name.blank? || Tutor::SubSystems.valid_name?(my_subsystem_name)
```
Comment (nathanstitt) [inline lib/tutor/subsystems/association_extensions.rb:41]: this seems safe but I'm not sure why the additional `my_subsystem_name.blank?` check is needed on line 40

File: app/routines/mark_task_step_completed.rb [hunk 4]
```diff
@@ -26,9 +26,37 @@ def exec(task_step:, completed_at: Time.current, lock_task: true, save: true)
...
 
+    queue = task.is_preview ? 'preview' : 'dashboard'
+    role_run_at = task.feedback_available? ? completed_at :
+                                             [ task.feedback_at, completed_at ].compact.max
+
+    page = Content::Models::Page
+      .select(:uuid, :parent_book_part_uuid)
+      .find_by(id: task_step.content_page_id)
+
+    unless page.nil?
```
[the remark this answers is not in the document]
Comment (Dantemss) [reply app/routines/mark_task_step_completed.rb:42]: Exists triggers another query though...

File: db/background_migrate/20200501125646_initialize_clue_caches.rb (hunk 1 above)
Comment (Dantemss) [reply db/background_migrate/20200501125646_initialize_clue_caches.rb:49]: Well it is a background migration... and we can leave the CLUes in the old mode until it finishes

File: db/migrate/20200428154625_add_parent_book_part_uuid_to_content_pages.rb (hunk 2 above)
Comment (Dantemss) [reply db/migrate/20200428154625_add_parent_book_part_uuid_to_content_pages.rb:11]: The units and chapters that were missing uuids had those added.

File: lib/tutor/subsystems/association_extensions.rb (hunk 3 above)
Comment (Dantemss) [reply lib/tutor/subsystems/association_extensions.rb:41]: I think you might be right, since the other method would probably return false.

Comment (Dantemss) [reply lib/tutor/subsystems/association_extensions.rb:41]: Ah no now I remember... I wanted to be able to use the `subsystem: :whatever` key to refer to subsystem stuff from stuff that was outside of subsystems (I ended up making a cache subsystem in the end though).

**Pull request merged**